In [1]:
import pandas as pd
import numpy as np

In [2]:
url = 'https://github.com/Nahorskyi/hackathon_INT20H_test/blob/main/int20h-ds-test-dataset.csv?raw=true'
df = pd.read_csv(url)
df=df.set_index('userid')
df.pop('event_created_date')
df=df.drop_duplicates(keep='first')
df


user_state                  event_name  \
userid                                                                    
c95c777785faec8dd910d019d7278ebe         CA         Add Vehicle Success   
c95c777785faec8dd910d019d7278ebe         CA           Add Vehicle Break   
f344be2d9a042b7444f3cc5279e38ef1         FL             Calculator View   
c95c777785faec8dd910d019d7278ebe         CA  Add Payment Method Success   
e331ed81422d8fba55520a43a872e701         IL             Sign Up Success   
...                                     ...                         ...   
59e61796833e7e9edcb7796fac1688ba         FL         Add Vehicle Success   
679eba26c4e75e0afb178360becfa21b         CA  Add Payment Method Success   
679eba26c4e75e0afb178360becfa21b         CA  Account Setup Profile Skip   
679eba26c4e75e0afb178360becfa21b         CA  Account Setup Profile Skip   
679eba26c4e75e0afb178360becfa21b         CA    Chat Conversation Opened   

                                                                   event_attributes  \
userid                                                                                
c95c777785faec8dd910d019d7278ebe  {"Make":"Dodge","Model":"Caravan","Color":"Whi...   
c95c777785faec8dd910d019d7278ebe                                                 {}   
f344be2d9a042b7444f3cc5279e38ef1                                                 {}   
c95c777785faec8dd910d019d7278ebe     {"Payment Method":"Credit","Tokenized Pay":""}   
e331ed81422d8fba55520a43a872e701                                 {"Method":"Apple"}   
...                                                                             ...   
59e61796833e7e9edcb7796fac1688ba  {"Year":"2015","Model":"Hino","Temporary":"fal...   
679eba26c4e75e0afb178360becfa21b  {"Payment Method":"Credit","Tokenized Pay":"",...   
679eba26c4e75e0afb178360becfa21b                               {"Screen":"Address"}   
679eba26c4e75e0afb178360becfa21b                          {"Screen":"Phone Number"}   
679eba26c4e75e0afb178360becfa21b         {"From":"Dashboard","Transaction type":""}   

                                 event_platform device_manufacture  \
userid                                                               
c95c777785faec8dd910d019d7278ebe        android            samsung   
c95c777785faec8dd910d019d7278ebe        android            samsung   
f344be2d9a042b7444f3cc5279e38ef1        android            samsung   
c95c777785faec8dd910d019d7278ebe        android            samsung   
e331ed81422d8fba55520a43a872e701            ios              Apple   
...                                         ...                ...   
59e61796833e7e9edcb7796fac1688ba            ios              Apple   
679eba26c4e75e0afb178360becfa21b        android             Google   
679eba26c4e75e0afb178360becfa21b        android             Google   
679eba26c4e75e0afb178360becfa21b        android             Google   
679eba26c4e75e0afb178360becfa21b        android             Google   

                                 device_model  
userid                                         
c95c777785faec8dd910d019d7278ebe     SM-N975U  
c95c777785faec8dd910d019d7278ebe     SM-N975U  
f344be2d9a042b7444f3cc5279e38ef1    SM-G973U1  
c95c777785faec8dd910d019d7278ebe     SM-N975U  
e331ed81422d8fba55520a43a872e701   iPhone12,1  
...                                       ...  
59e61796833e7e9edcb7796fac1688ba   iPhone13,1  
679eba26c4e75e0afb178360becfa21b     Pixel 3a  
679eba26c4e75e0afb178360becfa21b     Pixel 3a  
679eba26c4e75e0afb178360becfa21b     Pixel 3a  
679eba26c4e75e0afb178360becfa21b     Pixel 3a  

[8184 rows x 6 columns]

In [3]:
#Знаходимо унікальні значення для колонки event_name
df.event_name.unique()

array(['Add Vehicle Success', 'Add Vehicle Break', 'Calculator View',
       'Add Payment Method Success', 'Sign Up Success',
       'Add Payment Method Failed', 'Wallet Opened', 'Transaction Refund',
       'Email Confirmation Success', 'Account Setup Skip',
       'Subscription Premium', 'Chat Conversation Opened',
       'Account Setup Profile Skip',
       'Account History Transaction Details', 'Sign Out',
       'Chat Conversation Started', 'Order', 'Sign Up Error',
       'Subscription Premium Cancel', 'Calculator Used',
       'Subscription Premium Renew', 'Add Vehicle Failed',
       'Reset Password Set'], dtype=object)

З умови завдання нам потрібні ці івенти:
- Transaction Refund 
- Order 
- Account Setup Skip 
- Signup Success 
- Add Vehicle Success 
- Add Vehicle Break
- Add Payment Method Sucess 
- Add Payment Method Failed
- Email Confirmation Success
- Chat conversation Opened
- Sign Out 
- Account History Transaction Details
- Wallet Opened
- Subscription Premium 
- Subscription Premium Cancel
- Calculator view 

In [4]:
# Створюємо датафрейми для кожного івенту з колонки 'event_name' 
# з його параметрами з колонки 'event_attributes' 

Transaction_Refund         = df[df['event_name'] == 'Transaction Refund'][['event_attributes']] 
Transaction_Refund.rename(columns = {'event_attributes': 'Transaction_Refund'}, inplace=True)
Transaction_Refund=Transaction_Refund.groupby(by=["userid"]).sum()

Order = df[df['event_name'] == 'Order'][['event_attributes']] 
Order.rename(columns = {'event_attributes':'Order'} , inplace = True ) 
Order=Order.groupby(by=["userid"]).sum()

Account_Setup_Profile_Skip = df[df['event_name'] == 'Account Setup Profile Skip'][['event_attributes']] 
Account_Setup_Profile_Skip.rename(columns = {'event_attributes': 'Account_Setup_Profile_Skip'}, inplace=True)
Account_Setup_Profile_Skip=Account_Setup_Profile_Skip.groupby(by=["userid"]).sum()

Sign_Up_Success = df[df['event_name'] == 'Sign Up Success'][['event_attributes']] 
Sign_Up_Success.rename(columns = {'event_attributes': 'Sign_Up_Success'}, inplace=True)
Sign_Up_Success=Sign_Up_Success.groupby(by=["userid"]).sum()

Add_Vehicle_Success = df[df['event_name'] == 'Add Vehicle Success'][['event_attributes']] 
Add_Vehicle_Success.rename(columns = {'event_attributes': 'Add_Vehicle_Success'}, inplace=True) 
Add_Vehicle_Success=Add_Vehicle_Success.groupby(by=["userid"]).sum()

Add_Vehicle_Break  = df[df['event_name'] == 'Add Vehicle Break'][['event_attributes']] 
Add_Vehicle_Break.rename(columns = {'event_attributes': 'Add_Vehicle_Break'}, inplace=True)
Add_Vehicle_Break=Add_Vehicle_Break.groupby(by=["userid"]).sum()

Add_Payment_Method_Success = df[df['event_name'] == 'Add Payment Method Success'][['event_attributes']] 
Add_Payment_Method_Success.rename(columns = {'event_attributes': 'Add_Payment_Method_Success'}, inplace=True) 
Add_Payment_Method_Success=Add_Payment_Method_Success.groupby(by=["userid"]).sum()

Add_Payment_Method_Failed   = df[df['event_name'] == 'Add Payment Method Failed'][['event_attributes']] 
Add_Payment_Method_Failed.rename(columns = {'event_attributes': 'Add_Payment_Method_Failed'}, inplace=True)
Add_Payment_Method_Failed=Add_Payment_Method_Failed.groupby(by=["userid"]).sum()

Email_Confirmation_Success = df[df['event_name'] == 'Email Confirmation Success'][['event_attributes']] 
Email_Confirmation_Success.rename(columns = {'event_attributes': 'Email_Confirmation_Success'}, inplace=True)
Email_Confirmation_Success=Email_Confirmation_Success.groupby(by=["userid"]).sum()

Chat_Conversation_Opened   = df[df['event_name'] == 'Chat Conversation Opened'][['event_attributes']] 
Chat_Conversation_Opened.rename(columns = {'event_attributes': 'Chat_Conversation_Opened'}, inplace=True) 
Chat_Conversation_Opened=Chat_Conversation_Opened.groupby(by=["userid"]).sum()

Sign_Out = df[df['event_name'] == 'Sign Out'][['event_attributes']] 
Sign_Out.rename(columns = {'event_attributes': 'Sign_Out'}, inplace=True)
Sign_Out=Sign_Out.groupby(by=["userid"]).sum()

Account_History_Transaction_Details = df[df['event_name'] == 'Account History Transaction Details'][['event_attributes']] 
Account_History_Transaction_Details.rename(columns = {'event_attributes': 'Account_History_Transaction_Details'}, inplace=True) 
Account_History_Transaction_Details=Account_History_Transaction_Details.groupby(by=["userid"]).sum()

Wallet_Opened   = df[df['event_name'] == 'Wallet Opened'][['event_attributes']] 
Wallet_Opened.rename(columns = {'event_attributes': 'Wallet_Opened'}, inplace=True) 
Wallet_Opened=Wallet_Opened.groupby(by=["userid"]).sum()

Subscription_Premium = df[df['event_name'] == 'Subscription Premium'][['event_attributes']] 
Subscription_Premium.rename(columns = {'event_attributes': 'Subscription_Premium'}, inplace=True)
Subscription_Premium=Subscription_Premium.groupby(by=["userid"]).sum()

Subscription_Premium_Cancel   = df[df['event_name'] == 'Subscription Premium Cancel'][['event_attributes']] 
Subscription_Premium_Cancel.rename(columns = {'event_attributes': 'Subscription_Premium_Cancel'}, inplace=True)
Subscription_Premium_Cancel=Subscription_Premium_Cancel.groupby(by=["userid"]).sum()

Calculator_View = df[df['event_name'] == 'Calculator View'][['event_attributes']] 
Calculator_View.rename(columns = {'event_attributes': 'Calculator_View'}, inplace=True)
Calculator_View=Calculator_View.groupby(by=["userid"]).sum()

In [5]:
# Конкатенуємо датафрейми в один великий 
res = pd.concat([Sign_Up_Success,Order, Add_Vehicle_Success,Account_History_Transaction_Details, Chat_Conversation_Opened, Wallet_Opened, Add_Payment_Method_Success, Email_Confirmation_Success, Subscription_Premium, Calculator_View, Sign_Out, Account_Setup_Profile_Skip, Add_Vehicle_Break, Add_Payment_Method_Failed,  Transaction_Refund,Subscription_Premium_Cancel], axis=1, join='outer')
res

Sign_Up_Success Order  \
userid                                                        
001244c572f1a681553bc045a378cacf  {"Method":"Google"}   NaN   
00596aa4edee32815aaf4aad659c484e  {"Method":"Google"}   NaN   
0079e6902320d6771024a236ab801f2e  {"Method":"Google"}   NaN   
008dcdca320fbd55c289dd1f6d21d672  {"Method":"Google"}   NaN   
00db2bff1ee57b9d389312209bd40b4f  {"Method":"Google"}   NaN   
...                                               ...   ...   
21f7387a93418b77d33bfa56dcc0e3a9                  NaN   NaN   
b54bc83dfcefe27e4318c7dd0dab1f0a                  NaN   NaN   
c54556dbb037662501bf2f938e6c9fd4                  NaN   NaN   
c664cc762bdcb88230b4b4f07e3b7d2e                  NaN   NaN   
bb35dd1dfe0fcead9f5c5b0b3b9829dd                  NaN   NaN   

                                                                Add_Vehicle_Success  \
userid                                                                                
001244c572f1a681553bc045a378cacf                                                NaN   
00596aa4edee32815aaf4aad659c484e  {"Make":"Chrysler","Model":"300","Color":"Silv...   
0079e6902320d6771024a236ab801f2e                                                NaN   
008dcdca320fbd55c289dd1f6d21d672  {"Year":"2016","Model":"Mustang","Temporary":"...   
00db2bff1ee57b9d389312209bd40b4f                                                NaN   
...                                                                             ...   
21f7387a93418b77d33bfa56dcc0e3a9                                                NaN   
b54bc83dfcefe27e4318c7dd0dab1f0a                                                NaN   
c54556dbb037662501bf2f938e6c9fd4                                                NaN   
c664cc762bdcb88230b4b4f07e3b7d2e                                                NaN   
bb35dd1dfe0fcead9f5c5b0b3b9829dd                                                NaN   

                                 Account_History_Transaction_Details  \
userid                                                                 
001244c572f1a681553bc045a378cacf                                 NaN   
00596aa4edee32815aaf4aad659c484e                                 NaN   
0079e6902320d6771024a236ab801f2e                                 NaN   
008dcdca320fbd55c289dd1f6d21d672                                 NaN   
00db2bff1ee57b9d389312209bd40b4f                                 NaN   
...                                                              ...   
21f7387a93418b77d33bfa56dcc0e3a9                                 NaN   
b54bc83dfcefe27e4318c7dd0dab1f0a                                 NaN   
c54556dbb037662501bf2f938e6c9fd4                                 NaN   
c664cc762bdcb88230b4b4f07e3b7d2e                                 NaN   
bb35dd1dfe0fcead9f5c5b0b3b9829dd                                 NaN   

                                 Chat_Conversation_Opened Wallet_Opened  \
userid                                                                    
001244c572f1a681553bc045a378cacf                      NaN           NaN   
00596aa4edee32815aaf4aad659c484e                      NaN           NaN   
0079e6902320d6771024a236ab801f2e                      NaN           NaN   
008dcdca320fbd55c289dd1f6d21d672                      NaN            {}   
00db2bff1ee57b9d389312209bd40b4f                      NaN           NaN   
...                                                   ...           ...   
21f7387a93418b77d33bfa56dcc0e3a9                      NaN           NaN   
b54bc83dfcefe27e4318c7dd0dab1f0a                      NaN           NaN   
c54556dbb037662501bf2f938e6c9fd4                      NaN           NaN   
c664cc762bdcb88230b4b4f07e3b7d2e                      NaN           NaN   
bb35dd1dfe0fcead9f5c5b0b3b9829dd                      NaN           NaN   

                                                      Add_Payment_Method_Success  \
userid                                                                             
00124

In [6]:
# Замінюємо порожні дужки {} на 1.
# Порожні дужки використовувалися в деяких колонках задля 
# пояснення, що користувач виконував цю дію
res = res.replace(['{}'],'1')
# Nan замінюємо на 0, це означає, що користувач 
# івент з колонки не робив
res = res.fillna(0)
res

Sign_Up_Success Order  \
userid                                                        
001244c572f1a681553bc045a378cacf  {"Method":"Google"}     0   
00596aa4edee32815aaf4aad659c484e  {"Method":"Google"}     0   
0079e6902320d6771024a236ab801f2e  {"Method":"Google"}     0   
008dcdca320fbd55c289dd1f6d21d672  {"Method":"Google"}     0   
00db2bff1ee57b9d389312209bd40b4f  {"Method":"Google"}     0   
...                                               ...   ...   
21f7387a93418b77d33bfa56dcc0e3a9                    0     0   
b54bc83dfcefe27e4318c7dd0dab1f0a                    0     0   
c54556dbb037662501bf2f938e6c9fd4                    0     0   
c664cc762bdcb88230b4b4f07e3b7d2e                    0     0   
bb35dd1dfe0fcead9f5c5b0b3b9829dd                    0     0   

                                                                Add_Vehicle_Success  \
userid                                                                                
001244c572f1a681553bc045a378cacf                                                  0   
00596aa4edee32815aaf4aad659c484e  {"Make":"Chrysler","Model":"300","Color":"Silv...   
0079e6902320d6771024a236ab801f2e                                                  0   
008dcdca320fbd55c289dd1f6d21d672  {"Year":"2016","Model":"Mustang","Temporary":"...   
00db2bff1ee57b9d389312209bd40b4f                                                  0   
...                                                                             ...   
21f7387a93418b77d33bfa56dcc0e3a9                                                  0   
b54bc83dfcefe27e4318c7dd0dab1f0a                                                  0   
c54556dbb037662501bf2f938e6c9fd4                                                  0   
c664cc762bdcb88230b4b4f07e3b7d2e                                                  0   
bb35dd1dfe0fcead9f5c5b0b3b9829dd                                                  0   

                                 Account_History_Transaction_Details  \
userid                                                                 
001244c572f1a681553bc045a378cacf                                   0   
00596aa4edee32815aaf4aad659c484e                                   0   
0079e6902320d6771024a236ab801f2e                                   0   
008dcdca320fbd55c289dd1f6d21d672                                   0   
00db2bff1ee57b9d389312209bd40b4f                                   0   
...                                                              ...   
21f7387a93418b77d33bfa56dcc0e3a9                                   0   
b54bc83dfcefe27e4318c7dd0dab1f0a                                   0   
c54556dbb037662501bf2f938e6c9fd4                                   0   
c664cc762bdcb88230b4b4f07e3b7d2e                                   0   
bb35dd1dfe0fcead9f5c5b0b3b9829dd                                   0   

                                 Chat_Conversation_Opened Wallet_Opened  \
userid                                                                    
001244c572f1a681553bc045a378cacf                        0             0   
00596aa4edee32815aaf4aad659c484e                        0             0   
0079e6902320d6771024a236ab801f2e                        0             0   
008dcdca320fbd55c289dd1f6d21d672                        0             1   
00db2bff1ee57b9d389312209bd40b4f                        0             0   
...                                                   ...           ...   
21f7387a93418b77d33bfa56dcc0e3a9                        0             0   
b54bc83dfcefe27e4318c7dd0dab1f0a                        0             0   
c54556dbb037662501bf2f938e6c9fd4                        0             0   
c664cc762bdcb88230b4b4f07e3b7d2e                        0             0   
bb35dd1dfe0fcead9f5c5b0b3b9829dd                        0             0   

                                                      Add_Payment_Method_Success  \
userid                                                                             
00124

- Subscription_Premium_Cancel

In [7]:
Subscription_Premium_Cancel = Subscription_Premium_Cancel .replace(['{}'],'1')
Subscription_Premium_Cancel = Subscription_Premium_Cancel.fillna(0)
Subscription_Premium_Cancel.value_counts()

Subscription_Premium_Cancel
1                              26
dtype: int64

- Sign_Up_Success

In [8]:
Sign_Up_Success.Sign_Up_Success.value_counts()

{"Method":"Google"}      647
{"Method":"Email"}       491
{"Method":"Facebook"}    162
{"Method":"Apple"}       160
Name: Sign_Up_Success, dtype: int64

In [9]:
# Apple
Sign_Up_Success_Apple=Sign_Up_Success[Sign_Up_Success['Sign_Up_Success']=='{"Method":"Apple"}']
Sign_Up_Success_Apple[~Sign_Up_Success_Apple["Sign_Up_Success"].isnull()] = 1
Sign_Up_Success_Apple.rename(columns = {'Sign_Up_Success': 'Sign_Up_Success_Apple'}, inplace=True)


#Email
Sign_Up_Success_Email=Sign_Up_Success[Sign_Up_Success['Sign_Up_Success']=='{"Method":"Email"}']
Sign_Up_Success_Email.loc[~Sign_Up_Success_Email['Sign_Up_Success'].isnull()] = 1
Sign_Up_Success_Email.rename(columns = {'Sign_Up_Success': 'Sign_Up_Success_Email'}, inplace=True)


#Facebook
Sign_Up_Success_Facebook=Sign_Up_Success[Sign_Up_Success['Sign_Up_Success']=='{"Method":"Facebook"}']
Sign_Up_Success_Facebook.loc[~Sign_Up_Success_Facebook['Sign_Up_Success'].isnull()] = 1
Sign_Up_Success_Facebook.rename(columns = {'Sign_Up_Success': 'Sign_Up_Success_Facebook'}, inplace=True)


#Google
Sign_Up_Success_Google=Sign_Up_Success[Sign_Up_Success['Sign_Up_Success']=='{"Method":"Google"}']
Sign_Up_Success_Google.loc[~Sign_Up_Success_Google['Sign_Up_Success'].isnull()] = 1
Sign_Up_Success_Google.rename(columns = {'Sign_Up_Success': 'Sign_Up_Success_Google'}, inplace=True)

C:\Users\Max\AppData\Local\Temp\ipykernel_16360\2709548679.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  Sign_Up_Success_Apple[~Sign_Up_Success_Apple["Sign_Up_Success"].isnull()] = 1
C:\Users\Max\AppData\Local\Temp\ipykernel_16360\2709548679.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  Sign_Up_Success_Apple[~Sign_Up_Success_Apple["Sign_Up_Success"].isnull()] = 1
C:\Users\Max\AppData\Local\Temp\ipykernel_16360\2709548679.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentati

In [10]:
Sign_Up_Success_all = pd.concat([Sign_Up_Success_Apple,Sign_Up_Success_Email,Sign_Up_Success_Facebook,Sign_Up_Success_Google], axis=1, join='outer')
Sign_Up_Success_all = Sign_Up_Success_all.fillna(0)
Sign_Up_Success_all.astype(int)

Sign_Up_Success_Apple  \
userid                                                    
0105f45c4b5ba2d3a71b6b5a3115133b                      1   
04d80e22cc101455d8bbe476e34d6ac2                      1   
05a6df70d2d9e458be42c4a09c86c253                      1   
05bd68794e1c20075e4e42f0412bb94f                      1   
062965e6807c527aa818a7d670b5d935                      1   
...                                                 ...   
fc8c8752c6099c69f11d3ebd7738b3bf                      0   
fd5bc86aa11a1c4b1d17f1e78f50e4a5                      0   
fd91b5e00f7782750d8f3f0d37f6ece9                      0   
fe5dba6215aa96ba863ef6b9fd49c4eb                      0   
ff72157f48e62a7278d4efb45e876a4e                      0   

                                  Sign_Up_Success_Email  \
userid                                                    
0105f45c4b5ba2d3a71b6b5a3115133b                      0   
04d80e22cc101455d8bbe476e34d6ac2                      0   
05a6df70d2d9e458be42c4a09c86c253                      0   
05bd68794e1c20075e4e42f0412bb94f                      0   
062965e6807c527aa818a7d670b5d935                      0   
...                                                 ...   
fc8c8752c6099c69f11d3ebd7738b3bf                      0   
fd5bc86aa11a1c4b1d17f1e78f50e4a5                      0   
fd91b5e00f7782750d8f3f0d37f6ece9                      0   
fe5dba6215aa96ba863ef6b9fd49c4eb                      0   
ff72157f48e62a7278d4efb45e876a4e                      0   

                                  Sign_Up_Success_Facebook  \
userid                                                       
0105f45c4b5ba2d3a71b6b5a3115133b                         0   
04d80e22cc101455d8bbe476e34d6ac2                         0   
05a6df70d2d9e458be42c4a09c86c253                         0   
05bd68794e1c20075e4e42f0412bb94f                         0   
062965e6807c527aa818a7d670b5d935                         0   
...                                                    ...   
fc8c8752c6099c69f11d3ebd7738b3bf                         0   
fd5bc86aa11a1c4b1d17f1e78f50e4a5                         0   
fd91b5e00f7782750d8f3f0d37f6ece9                         0   
fe5dba6215aa96ba863ef6b9fd49c4eb                         0   
ff72157f48e62a7278d4efb45e876a4e                         0   

                                  Sign_Up_Success_Google  
userid                                                    
0105f45c4b5ba2d3a71b6b5a3115133b                       0  
04d80e22cc101455d8bbe476e34d6ac2                       0  
05a6df70d2d9e458be42c4a09c86c253                       0  
05bd68794e1c20075e4e42f0412bb94f                       0  
062965e6807c527aa818a7d670b5d935                       0  
...                                                  ...  
fc8c8752c6099c69f11d3ebd7738b3bf                       1  
fd5bc86aa11a1c4b1d17f1e78f50e4a5                       1  
fd91b5e00f7782750d8f3f0d37f6ece9                       1  
fe5dba6215aa96ba863ef6b9fd49c4eb                       1  
ff72157f48e62a7278d4efb45e876a4e                       1  

[1460 rows x 4 columns]

In [11]:
Sign_Up_Success_all_SPC = pd.concat([Sign_Up_Success_all, Subscription_Premium_Cancel], axis=1, join='outer')
Sign_Up_Success_all_SPC = Sign_Up_Success_all_SPC.fillna(0)
Sign_Up_Success_all_SPC=Sign_Up_Success_all_SPC.astype(int)
Sign_Up_Success_all_SPC_corr=Sign_Up_Success_all_SPC.corr()
Sign_Up_Success_all_SPC_corr=Sign_Up_Success_all_SPC_corr.iloc[-1]
Sign_Up_Success_all_SPC_corr.sort_values(ascending = False)

Subscription_Premium_Cancel    1.000000
Sign_Up_Success_Apple         -0.013753
Sign_Up_Success_Email         -0.018431
Sign_Up_Success_Google        -0.046198
Sign_Up_Success_Facebook      -0.047223
Name: Subscription_Premium_Cancel, dtype: float64

Бачимо, що немає залежності між параметрами Sign_Up_Success (Apple, Email, Google, Facebook) та Subscription_Premium_Cancel

- Order

In [12]:
Order.Order.value_counts()

{"Premium Membership":"True"}     27
{"Premium Membership":"False"}    12
Name: Order, dtype: int64

In [13]:
#True
Order_True=Order[Order['Order']=='{"Premium Membership":"True"}']
Order_True[~Order_True["Order"].isnull()] = 1
Order_True.rename(columns = {'Order': 'Order_True'}, inplace=True)

#False
Order_False=Order[Order['Order']=='{"Premium Membership":"False"}']
Order_False[~Order_False["Order"].isnull()] = 1
Order_False.rename(columns = {'Order': 'Order_False'}, inplace=True)

C:\Users\Max\AppData\Local\Temp\ipykernel_16360\1516560079.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  Order_True[~Order_True["Order"].isnull()] = 1
C:\Users\Max\AppData\Local\Temp\ipykernel_16360\1516560079.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  Order_True[~Order_True["Order"].isnull()] = 1
C:\Users\Max\AppData\Local\Temp\ipykernel_16360\1516560079.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/inde

In [14]:
Order_all = pd.concat([Order_True,Order_False], axis=1, join='outer')
Order_all = Order_all.fillna(0)
Order_all.astype(int)

Order_True  Order_False
userid                                                   
0d51394c39dbf15021ccd09155ee6257           1            0
1c18dbe71e395e81a4c2f1640065f2df           1            0
1c5d42f9cf94d9eca90b3628d4d7f75d           1            0
28368284a145474b2cc4961349bffcf3           1            0
30560b6ff7feef07c9cb7315b0564101           1            0
33d4c1a149926acd6536452ecbddc8e2           1            0
44cd60c0ddb41223eb13c1a224b5e3d6           1            0
4949c7b6dfa8cef23d641fd98d50ab18           1            0
66d1fab17f5b7bcf9e43ae0438a2a6cf           1            0
746355602a97e0c70015a35f67a96f71           1            0
8f2cf06465717d43a9e467baaf510feb           1            0
a23a2974a70882b4bd66da405a2b80a5           1            0
a2a8a4e98c233aa86d5b34de7ba6f1d3           1            0
a36f9c3e8bcc3c6ac7ee75f9e08c1deb           1            0
ae932c52b76c8deb61f8555f64d38490           1            0
b58466f0302ba0e51063a9711f153e51           1            0
c0ce0b06f849d0a75dc3f2e69c87b0b6           1            0
c88aebef8d6337c3ab4c04be6853976a           1            0
d3d51e2fd28f62d02e93f3aae6179246           1            0
d97569a26b47b089e4ab0bce7012a12e           1            0
dce57d89ddc943bb268e3a2bcd8b91c0           1            0
ddfd6a4298e15df738efd4daa76403e4           1            0
eba4e8c1c66ee2e1b54ae61f0ae16190           1            0
f24dd26af8643f2a491a01f7e25051da           1            0
f40765f0ec0f11bbfdf99a564ce9f86d           1            0
f71d97f27525dd75106f9c593e31abf4           1            0
fdb8c0dd34b05d1f1b0eb6f1113df175           1            0
26cc023b7271834292da520c97fbef3d           0            1
3470f4afabdbf091c92349cd59b87962           0            1
6913af0efb0e6d62a4ed4ebd949f772e           0            1
90e7a324a520fc15e8acec664e9996ab           0            1
9c4b8ec536508a6c1cf9ab66f4200589           0            1
b10e7150d93e6ef244286da36fb38220           0            1
b8e494af62901281ea32197a58a5fdc7           0            1
bf6f5c66e2438576cc4852345313fe55           0            1
d1e49bffcac66079c0ffa353ef660c87           0            1
da115da0c38761649740d23199499d0f           0            1
da823e0efb16f00e73e2cbf1d8951649           0            1
dce56a898f7c104b27cd482b8f95da46           0            1

In [15]:
Order_all_SPC = pd.concat([Order_all, Subscription_Premium_Cancel], axis=1, join='outer')
Order_all_SPC = Order_all_SPC.fillna(0)
Order_all_SPC=Order_all_SPC.astype(int)
Order_all_SPC_corr=Order_all_SPC.corr()
Order_all_SPC_corr=Order_all_SPC_corr.iloc[-1]
Order_all_SPC_corr.sort_values(ascending = False)

Subscription_Premium_Cancel    1.000000
Order_True                    -0.404184
Order_False                   -0.448509
Name: Subscription_Premium_Cancel, dtype: float64

Бачимо, що немає залежності між параметрами Order (Order_True , Order_False) та Subscription_Premium_Cancel

- Add_Vehicle_Succes

In [16]:
Add_Vehicle_Success.value_counts

<bound method DataFrame.value_counts of                                                                 Add_Vehicle_Success
userid                                                                             
0006869712ec9841dc36234bce245203  {"Make":"Chevrolet","Model":"Express 3500","Co...
003f57fe2631ade57a86f6a2b96bb20c  {"Make":"Ford","Model":"C-Max Energi","Tempora...
00596aa4edee32815aaf4aad659c484e  {"Make":"Chrysler","Model":"300","Color":"Silv...
007b54670d72df309afd87befcbb63db  {"Start Date":"05/04/2022","Year":"2022","Mode...
008dcdca320fbd55c289dd1f6d21d672  {"Year":"2016","Model":"Mustang","Temporary":"...
...                                                                             ...
feeb75def9523b6c8c1d804267988c0d  {"Make":"Buick","Model":"LaCrosse","Temporary"...
ff189ff549ef3613dd2f275a82c0bbf8  {"Year":"2021","Model":"Highlander Hybrid","Te...
ff530883fc9c83ac50418d8672e9056e  {"Year":"2016","Model":"CX-5","Temporary":"fal...
ff72157f48e62a7278d4efb45e876a4e  {"

In [17]:
list_car =[ "Abarth", "Alfa Romeo", "Aston Martin", "Audi", "Bentley", "BMW", "Bugatti", "Cadillac", "Chevrolet", "Chrysler", "Citroën", "Dacia", "Daewoo", "Daihatsu", "Dodge", "Donkervoort", "DS", "Ferrari", "Fiat", "Fisker", "Ford", "Honda", "Hummer", "Hyundai", "Infiniti", "Iveco", "Jaguar", "Jeep", "Kia", "KTM", "Lada", "Lamborghini", "Lancia", "Land Rover", "Landwind", "Lexus", "Lotus", "Maserati", "Maybach", "Mazda", "McLaren", "Mercedes-Benz", "MG", "Mini", "Mitsubishi", "Morgan", "Nissan", "Opel", "Peugeot", "Porsche", "Renault", "Rolls-Royce", "Rover", "Saab", "Seat", "Skoda", "Smart", "SsangYong", "Subaru", "Suzuki", "Tesla", "Toyota", "Volkswagen", "Volvo" ]

In [18]:
for car in list_car:
     Add_Vehicle_Success[car] = Add_Vehicle_Success['Add_Vehicle_Success'].str.find(car)

Add_Vehicle_Success=Add_Vehicle_Success.drop('Add_Vehicle_Success', axis=1)
Add_Vehicle_Success = Add_Vehicle_Success.replace(-1,0) 

for col in Add_Vehicle_Success:
    Add_Vehicle_Success[col].loc[Add_Vehicle_Success[col]>0] = 1

Add_Vehicle_Success = Add_Vehicle_Success.loc[:, (Add_Vehicle_Success != 0).any(axis=0)]    

Add_Vehicle_Success_all = Add_Vehicle_Success

In [19]:
Add_Vehicle_Success_SPC = pd.concat([Add_Vehicle_Success, Subscription_Premium_Cancel], axis=1, join='outer')
Add_Vehicle_Success_SPC = Add_Vehicle_Success_SPC.fillna(0)
Add_Vehicle_Success_SPC=Add_Vehicle_Success_SPC.astype(int)
Add_Vehicle_Success_SPC_corr=Add_Vehicle_Success_SPC.corr()
Add_Vehicle_Success_SPC_corr=Add_Vehicle_Success_SPC_corr.iloc[-1]
Add_Vehicle_Success_SPC_corr.sort_values(ascending = False)

Subscription_Premium_Cancel    1.000000
Mitsubishi                     0.042296
Audi                           0.038155
Honda                          0.012658
BMW                            0.011010
Kia                           -0.000584
Dodge                         -0.002554
Saab                          -0.002851
Alfa Romeo                    -0.002851
Jaguar                        -0.004033
Fiat                          -0.004033
Suzuki                        -0.005707
Land Rover                    -0.006993
Rover                         -0.006993
Volvo                         -0.008080
Porsche                       -0.008572
Chevrolet                     -0.008660
Mini                          -0.009038
Tesla                         -0.009038
Infiniti                      -0.009907
Cadillac                      -0.011085
Chrysler                      -0.013451
Mazda                         -0.015474
Subaru                        -0.016779
Lexus                         -0.017759


Бачимо, що маємо кореляцію між Add_Vehicle_Success (Mitsubishi) та Subscription_Premium_Cancel, яка дорівнює 0.042296

- Account_History_Transaction_Details

In [20]:
Account_History_Transaction_Details.value_counts

<bound method DataFrame.value_counts of                                                 Account_History_Transaction_Details
userid                                                                             
0006869712ec9841dc36234bce245203  {"Navigated From":"Account History","Status":"...
03e0c91e1163e8b80e74e586a3e666d6  {"Type":"Pro Plan Subscription","Navigated Fro...
0424df1cbc0b11f32e626f43fc176a4d  {"Navigated From":"Account History","Status":"...
05630c0e889fe7371fc74d7ae01f491f  {"Type":"Pro Plan Subscription","Navigated Fro...
060153cf7a5844d9aed735d408b35f97  {"Type":"Auto Payment","Navigated From":"accou...
...                                                                             ...
fc817111904f14d8868601ac1e1fedeb  {"Navigated From":"Account History","Status":"...
fc8cd52c073c56ad5f635f44097de339  {"Type":"Auto Payment","Navigated From":"accou...
fc8e16ce7c98f58ba309cb7fc3cd1dec  {"Type":"Pro Plan Subscription","Navigated Fro...
fca687b7a7b16e217c0a03b9ca8fdfda  {"

In [21]:
# Якщо клієнт відкривав історію акаунта - то 1, якщо не відкривав - то 0
Account_History_Transaction_Details.loc[~Account_History_Transaction_Details['Account_History_Transaction_Details'].isnull()] = 1

In [22]:
Account_History_Transaction_Details_SPC = pd.concat([Account_History_Transaction_Details, Subscription_Premium_Cancel], axis=1, join='outer')
Account_History_Transaction_Details_SPC = Account_History_Transaction_Details_SPC.fillna(0)
Account_History_Transaction_Details_SPC=Account_History_Transaction_Details_SPC.astype(int)
Account_History_Transaction_Details_SPC_corr=Account_History_Transaction_Details_SPC.corr()
Account_History_Transaction_Details_SPC_corr=Account_History_Transaction_Details_SPC_corr.iloc[-1]
Account_History_Transaction_Details_SPC_corr.sort_values(ascending = False)

Subscription_Premium_Cancel            1.000000
Account_History_Transaction_Details   -0.692549
Name: Subscription_Premium_Cancel, dtype: float64

Бачимо, що немає ніякої кореляції між тим, що клієнт дивився свою історію транзакції та тим, що клієнт скасував підписку

- Chat_Conversation_Opened

In [23]:
Chat_Conversation_Opened.value_counts

<bound method DataFrame.value_counts of                                                            Chat_Conversation_Opened
userid                                                                             
007b54670d72df309afd87befcbb63db                               {"From":"Dashboard"}
017f4557810b3ba165d2959c17b916b6                                    {"From":"Menu"}
02ea6c406a8d4d65cf25ea527ec83403              {"From":"Menu","Transaction type":""}
05d8b7e9ab7759e6d9d210aee9003379                  {"From":"","Transaction type":""}
078ae59f1ae79df514e846e5e51eca8b                                {"From":"Deeplink"}
...                                                                             ...
fa218673a8e63669d6839eeaebaca99a  {"From":"Transaction Details","Transaction typ...
fb8dec009d03ea9b6b7d385946f68643                {"From":"Dashboard"}{"From":"Menu"}
fc817111904f14d8868601ac1e1fedeb  {"From":"Transaction Details","Transaction typ...
fc8e16ce7c98f58ba309cb7fc3cd1dec  {"

In [24]:
# Якщо чат відкритий клієнтом - то 1, якщо ні - то 0
Chat_Conversation_Opened.loc[~Chat_Conversation_Opened['Chat_Conversation_Opened'].isnull()] = 1

In [25]:
Chat_Conversation_Opened_SPC = pd.concat([Chat_Conversation_Opened, Subscription_Premium_Cancel], axis=1, join='outer')
Chat_Conversation_Opened_SPC = Chat_Conversation_Opened_SPC.fillna(0)
Chat_Conversation_Opened_SPC=Chat_Conversation_Opened_SPC.astype(int)
Chat_Conversation_Opened_SPC_corr=Chat_Conversation_Opened_SPC.corr()
Chat_Conversation_Opened_SPC_corr=Chat_Conversation_Opened_SPC_corr.iloc[-1]
Chat_Conversation_Opened_SPC_corr.sort_values(ascending = False)

Subscription_Premium_Cancel    1.000000
Chat_Conversation_Opened      -0.664339
Name: Subscription_Premium_Cancel, dtype: float64

Бачимо, що немає ніякої кореляції між тим, що клієнт відкрив чат і тим, що клієнт скасував підписку

- Wallet_Opened

In [26]:
Wallet_Opened.value_counts()

Wallet_Opened
{}               263
{}{}               4
dtype: int64

{} - відкрив клієнт гаманець один раз 

{}{} - відкрив клієнт гаманець два рази

In [27]:
#1 раз
Wallet_Opened_1=Wallet_Opened[Wallet_Opened['Wallet_Opened']=='{}']
Wallet_Opened_1 = Wallet_Opened_1.replace(['{}'],1)
Wallet_Opened_1.rename(columns = {'Wallet_Opened': 'Wallet_Opened_1'}, inplace=True)

#2 рази
Wallet_Opened_2=Wallet_Opened[Wallet_Opened['Wallet_Opened']=='{}{}']
Wallet_Opened_2 = Wallet_Opened_2.replace(['{}{}'],1)
Wallet_Opened_2.rename(columns = {'Wallet_Opened': 'Wallet_Opened_2'}, inplace=True)

In [28]:
Wallet_Opened_all=pd.concat([Wallet_Opened_1,Wallet_Opened_2], axis=1)
Wallet_Opened_all = Wallet_Opened_all.fillna(0)
Wallet_Opened_all=Wallet_Opened_all.astype(int)

In [29]:
Wallet_Opened_all_SPC = pd.concat([Wallet_Opened_all, Subscription_Premium_Cancel], axis=1, join='outer')
Wallet_Opened_all_SPC = Wallet_Opened_all_SPC.fillna(0)
Wallet_Opened_all_SPC=Wallet_Opened_all_SPC.astype(int)
Wallet_Opened_all_SPC_corr=Wallet_Opened_all_SPC.corr()
Wallet_Opened_all_SPC_corr=Wallet_Opened_all_SPC_corr.iloc[-1]
Wallet_Opened_all_SPC_corr.sort_values(ascending = False)

Subscription_Premium_Cancel    1.000000
Wallet_Opened_2                0.064159
Wallet_Opened_1               -0.359297
Name: Subscription_Premium_Cancel, dtype: float64

Бачимо кореляцію між тим, що клієнт відкривав 2 рази гаманець та тим, що клієнт скасував підписку, вона дорівнює 0.064159

- Add_Payment_Method_Success

In [30]:
Add_Payment_Method_Success.value_counts

<bound method DataFrame.value_counts of                                                          Add_Payment_Method_Success
userid                                                                             
0006869712ec9841dc36234bce245203      {"Payment Method":"Debit","Tokenized Pay":""}
00596aa4edee32815aaf4aad659c484e     {"Payment Method":"PayPal","Tokenized Pay":""}
007b54670d72df309afd87befcbb63db         {"Promo code":"","Payment Method":"Debit"}
0136a8eeca5d16ac1a521dd0e27ecec0  {"Tokenized Pay":"Apple Pay","Payment Method":...
0136e4685e98e034bfbee6016680c21b  {"Payment Method":"Credit","Tokenized Pay":"Go...
...                                                                             ...
fd5bc86aa11a1c4b1d17f1e78f50e4a5     {"Payment Method":"PayPal","Tokenized Pay":""}
fdb8c0dd34b05d1f1b0eb6f1113df175  {"Tokenized Pay":"Apple Pay","Payment Method":...
fdb9c5f6193e1cc735a2c59e14882b0f  {"Payment Method":"Debit","Tokenized Pay":"","...
fdf6855262649124354ecf8c23f31356    

In [31]:
# Якщо клієнт успішно додавав метод оплати - то 1, якщо ні - то 0
Add_Payment_Method_Success.loc[~Add_Payment_Method_Success['Add_Payment_Method_Success'].isnull()] = 1
Add_Payment_Method_Success

Add_Payment_Method_Success
userid                                                     
0006869712ec9841dc36234bce245203                          1
00596aa4edee32815aaf4aad659c484e                          1
007b54670d72df309afd87befcbb63db                          1
0136a8eeca5d16ac1a521dd0e27ecec0                          1
0136e4685e98e034bfbee6016680c21b                          1
...                                                     ...
fd5bc86aa11a1c4b1d17f1e78f50e4a5                          1
fdb8c0dd34b05d1f1b0eb6f1113df175                          1
fdb9c5f6193e1cc735a2c59e14882b0f                          1
fdf6855262649124354ecf8c23f31356                          1
fffec8d843de9e49ef7c13df5a44da24                          1

[709 rows x 1 columns]

In [32]:
Add_Payment_Method_Success_SPC = pd.concat([Add_Payment_Method_Success, Subscription_Premium_Cancel], axis=1, join='outer')
Add_Payment_Method_Success_SPC = Add_Payment_Method_Success_SPC.fillna(0)
Add_Payment_Method_Success_SPC=Add_Payment_Method_Success_SPC.astype(int)
Add_Payment_Method_Success_SPC_corr=Add_Payment_Method_Success_SPC.corr()
Add_Payment_Method_Success_SPC_corr=Add_Payment_Method_Success_SPC_corr.iloc[-1]
Add_Payment_Method_Success_SPC_corr.sort_values(ascending = False)

Subscription_Premium_Cancel    1.000000
Add_Payment_Method_Success    -0.581252
Name: Subscription_Premium_Cancel, dtype: float64

Бачимо, що немає ніякої кореляції між тим, що успішно додавав метод оплати та тим, що він скасував підписку

- Email_Confirmation_Success

In [33]:
Email_Confirmation_Success.value_counts()

Email_Confirmation_Success
{}                            195
{"Required":"true"}           166
dtype: int64

In [34]:
# пошта підтверджена 
Email_Confirmation_Success_True=Email_Confirmation_Success[Email_Confirmation_Success['Email_Confirmation_Success']=='{}']
Email_Confirmation_Success_True = Email_Confirmation_Success_True.replace(['{}'],1)
Email_Confirmation_Success_True.rename(columns = {'Email_Confirmation_Success': 'Email_Confirmation_Success_True'}, inplace=True)

# потрібне підтвердження пошти
Email_Confirmation_Success_RT=Email_Confirmation_Success[Email_Confirmation_Success['Email_Confirmation_Success']=='{"Required":"true"}']
Email_Confirmation_Success_RT = Email_Confirmation_Success_RT.replace(['{"Required":"true"}'],1)
Email_Confirmation_Success_RT.rename(columns = {'Wallet_Opened': 'Wallet_Opened_RT'}, inplace=True)

In [35]:
Email_Confirmation_Success_all = pd.concat([Email_Confirmation_Success_True, Email_Confirmation_Success_RT], axis=1, join='outer')
Email_Confirmation_Success_all=Email_Confirmation_Success_all.fillna(0)

In [36]:
Email_Confirmation_Success_all_SPC = pd.concat([Email_Confirmation_Success_all, Subscription_Premium_Cancel], axis=1, join='outer')
Email_Confirmation_Success_all_SPC = Email_Confirmation_Success_all_SPC.fillna(0)
Email_Confirmation_Success_all_SPC= Email_Confirmation_Success_all_SPC.astype(int)
Email_Confirmation_Success_all_SPC_corr= Email_Confirmation_Success_all_SPC.corr()
Email_Confirmation_Success_all_SPC_corr=Email_Confirmation_Success_all_SPC_corr.iloc[-1]
Email_Confirmation_Success_all_SPC_corr.sort_values(ascending = False)

Subscription_Premium_Cancel        1.000000
Email_Confirmation_Success        -0.154628
Email_Confirmation_Success_True   -0.215676
Name: Subscription_Premium_Cancel, dtype: float64

Бачимо, що немає ніякої кореляції між тим, що клієнт почту підтвердив або не підтвердив та тим, що він скасував підписку

- Subscription_Premium

In [37]:
Subscription_Premium.value_counts()

Subscription_Premium                                                               
{"Start Date":"2022-03-16","Renewal Date":"2022-04-17","Membership Type":"Monthly"}    9
{"Start Date":"2022-03-26","Renewal Date":"2022-04-27","Membership Type":"Monthly"}    8
{"Start Date":"2022-04-15","Renewal Date":"2022-05-16","Membership Type":"Monthly"}    8
{"Start Date":"2022-03-02","Renewal Date":"2022-04-03","Membership Type":"Monthly"}    7
{"Start Date":"2022-02-19","Renewal Date":"2022-03-20","Membership Type":"Monthly"}    7
                                                                                      ..
{"Start Date":"2022-02-07","Renewal Date":"2022-03-08","Membership Type":"Monthly"}    2
{"Start Date":"2022-01-28","Renewal Date":"2022-03-01","Membership Type":"Monthly"}    1
{"Start Date":"2022-02-22","Renewal Date":"2022-03-23","Membership Type":"Monthly"}    1
{"Start Date":"2022-02-06","Renewal Date":"2022-03-07","Membership Type":"Monthly"}    1
{"Start Date":"2022-02-10"

In [38]:
# Якщо у клієнта є підписка - то 1, якщо ні - то 0
Subscription_Premium.loc[~Subscription_Premium['Subscription_Premium'].isnull()] = 1

In [39]:
Subscription_Premium_all_SPC = pd.concat([Subscription_Premium, Subscription_Premium_Cancel], axis=1, join='outer')
Subscription_Premium_all_SPC = Subscription_Premium_all_SPC.fillna(0)
Subscription_Premium_all_SPC=  Subscription_Premium_all_SPC.astype(int)
Subscription_Premium_all_SPC_corr=  Subscription_Premium_all_SPC.corr()
Subscription_Premium_all_SPC_corr= Subscription_Premium_all_SPC_corr.iloc[-1]
Subscription_Premium_all_SPC_corr.sort_values(ascending = False)

Subscription_Premium_Cancel    1.000000
Subscription_Premium          -0.606077
Name: Subscription_Premium_Cancel, dtype: float64

Бачимо, що немає кореляції між тим, чи є у клієнта підписка з тим, що він її скасував

- Calculator_View

In [40]:
Calculator_View.value_counts()

Calculator_View
{}                 231
{}{}                 2
dtype: int64

{} - відкрив клієнт калькулятор один раз 

{}{} - відкрив клієнт калькулятор два рази

In [41]:
#1 раз
Calculator_View_1=Calculator_View[Calculator_View['Calculator_View']=='{}']
Calculator_View_1 = Calculator_View_1.replace(['{}'],1)
Calculator_View_1.rename(columns = {'Calculator_View': 'Calculator_View_1'}, inplace=True)

#2 рази
Calculator_View_2=Calculator_View[Calculator_View['Calculator_View']=='{}{}']
Calculator_View_2 = Calculator_View_2.replace(['{}{}'],1)
Calculator_View_2.rename(columns = {'Calculator_View': 'Calculator_View_2'}, inplace=True)

In [42]:
Calculator_View_all=pd.concat([Calculator_View_1,Calculator_View_2], axis=1)
Calculator_View_all = Calculator_View_all.fillna(0)
Calculator_View_all=Calculator_View_all.astype(int)

In [43]:
Calculator_View_all_SPC = pd.concat([Calculator_View_all, Subscription_Premium_Cancel], axis=1, join='outer')
Calculator_View_all_SPC = Calculator_View_all_SPC.fillna(0)
Calculator_View_all_SPC=Calculator_View_all_SPC.astype(int)
Calculator_View_all_SPC_corr=Calculator_View_all_SPC.corr()
Calculator_View_all_SPC_corr=Calculator_View_all_SPC_corr.iloc[-1]
Calculator_View_all_SPC_corr.sort_values(ascending = False)

Subscription_Premium_Cancel    1.000000
Calculator_View_2              0.116774
Calculator_View_1             -0.865671
Name: Subscription_Premium_Cancel, dtype: float64

Бачимо кореляцію між тим, що клієнт відкрив калькулятор два рази з тим, що він скасував підписку і вона дорівнює 0.116774

- Sign_Out

In [44]:
Sign_Out.value_counts()

Sign_Out                          
{"Type":"Manual"}                     187
{"Type":"Auto"}                        35
{"Type":"Manual"}{"Type":"Manual"}      3
{"Type":"Auto"}{"Type":"Auto"}          1
{"Type":"Auto"}{"Type":"Manual"}        1
{"Type":"Manual"}{"Type":"Auto"}        1
dtype: int64

In [45]:
# Якщо клієнт вийшов - то 1, якщо ні - то 0
Sign_Out.loc[~Sign_Out['Sign_Out'].isnull()] = 1
Sign_Out

Sign_Out
userid                                   
0079e6902320d6771024a236ab801f2e        1
0136e4685e98e034bfbee6016680c21b        1
052642d5d16c869e7c341f53bd824328        1
05d8b7e9ab7759e6d9d210aee9003379        1
07a0d0ab07446da83579be3462bf081e        1
...                                   ...
fb6642c0e2ee8b65e04c4f0bc2c48d42        1
fc817111904f14d8868601ac1e1fedeb        1
fc8c8752c6099c69f11d3ebd7738b3bf        1
fd5bc86aa11a1c4b1d17f1e78f50e4a5        1
fdf6855262649124354ecf8c23f31356        1

[228 rows x 1 columns]

In [46]:
Sign_Out_SPC = pd.concat([Sign_Out, Subscription_Premium_Cancel], axis=1, join='outer')
Sign_Out_SPC = Sign_Out_SPC.fillna(0)
Sign_Out_SPC=Sign_Out_SPC.astype(int)
Sign_Out_SPC_corr=Sign_Out_SPC.corr()
Sign_Out_SPC_corr=Sign_Out_SPC_corr.iloc[-1]
Sign_Out_SPC_corr.sort_values(ascending = False)

Subscription_Premium_Cancel    1.000000
Sign_Out                      -0.865441
Name: Subscription_Premium_Cancel, dtype: float64

Маємо обернену кореляцію  між тим , що клієнт вийшов з акаунту з тим , що він скасував підписку, ця кореляція дорівнює -0.865441

- Account_Setup_Profile_Skip

In [47]:
Account_Setup_Profile_Skip.value_counts()

Account_Setup_Profile_Skip                                    
{"Screen":"Address"}{"Screen":"Phone number"}                     40
{"Screen":"Address"}                                              37
{"Screen":"Address"}{"Screen":"Phone Number"}                     32
{"Screen":"Address details"}                                      22
{"Screen":"Phone number"}                                         13
{"Screen":"Name"}                                                 10
{"Screen":"Name"}{"Screen":"Address"}{"Screen":"Phone Number"}     9
{"Screen":"Name"}{"Screen":"Address"}{"Screen":"Phone number"}     9
{"Screen":"Phone Number"}                                          9
{"Screen":"Address Details"}                                       3
{"Screen":"Address details"}{"Screen":"Phone number"}              2
{"Screen":"Phone number exists"}                                   2
dtype: int64

In [48]:
# Якщо клієнт пропустив налаштування профілю - то 1, якщо ні - то 0
Account_Setup_Profile_Skip.loc[~Account_Setup_Profile_Skip['Account_Setup_Profile_Skip'].isnull()] = 1
Account_Setup_Profile_Skip

Account_Setup_Profile_Skip
userid                                                     
0136e4685e98e034bfbee6016680c21b                          1
03a8fa7b862ce82bbc5963dc72e99e62                          1
052642d5d16c869e7c341f53bd824328                          1
0863b355ac677a9b85ce5984bf876dc5                          1
0a50cee56268e3c8875fcc59baaf3ee2                          1
...                                                     ...
f56ebcd64fc631a63524044081c72aeb                          1
f5b89fd61ad0cc105f6f0c48289c339a                          1
f5bd3257ad0018bf2715414b3272f840                          1
f87767cfda7bfd30820f0a53fd179eaa                          1
fa8574e4b97ff04270a538ab8d41a287                          1

[188 rows x 1 columns]

In [49]:
Account_Setup_Profile_Skip_SPC = pd.concat([Account_Setup_Profile_Skip, Subscription_Premium_Cancel], axis=1, join='outer')
Account_Setup_Profile_Skip_SPC = Account_Setup_Profile_Skip_SPC.fillna(0)
Account_Setup_Profile_Skip_SPC=Account_Setup_Profile_Skip_SPC.astype(int)
Account_Setup_Profile_Skip_SPC_corr=Account_Setup_Profile_Skip_SPC.corr()
Account_Setup_Profile_Skip_SPC_corr=Account_Setup_Profile_Skip_SPC_corr.iloc[-1]
Account_Setup_Profile_Skip_SPC_corr.sort_values(ascending = False)

Subscription_Premium_Cancel    1.000000
Account_Setup_Profile_Skip    -0.977969
Name: Subscription_Premium_Cancel, dtype: float64

Бачимо, що немає кореляції між клієнтами, які пропустили налаштування профілю з тим, що клієнти скасували підписку.
Але якщо клієнт не пропустив налаштування профілю, то видно, що в нас велика кореляція з тим, що він скасує підписку 

- Add_Vehicle_Break

In [50]:
Add_Vehicle_Break.value_counts()

Add_Vehicle_Break
{}                   211
dtype: int64

In [51]:
# Якщо клієнт зробив перерву в русі автомобіля - то 1, якщо ні - то 0
Add_Vehicle_Break.loc[~Add_Vehicle_Break['Add_Vehicle_Break'].isnull()] = 1
Add_Vehicle_Break

Add_Vehicle_Break
userid                                            
008dcdca320fbd55c289dd1f6d21d672                 1
01529f53aaba5b512168d432d5650556                 1
02ea6c406a8d4d65cf25ea527ec83403                 1
03a8fa7b862ce82bbc5963dc72e99e62                 1
03b74f9eedd297ea3dee1db83269d973                 1
...                                            ...
f40765f0ec0f11bbfdf99a564ce9f86d                 1
f7d597781ec1c52eea0c04de2269f1bc                 1
f81c03566135693f04d9f03016a5cc82                 1
f87767cfda7bfd30820f0a53fd179eaa                 1
f9ae02749af27f1aa62252d501c52fd1                 1

[211 rows x 1 columns]

In [52]:
Add_Vehicle_Break_SPC = pd.concat([Add_Vehicle_Break, Subscription_Premium_Cancel], axis=1, join='outer')
Add_Vehicle_Break_SPC = Add_Vehicle_Break_SPC.fillna(0)
Add_Vehicle_Break_SPC=Add_Vehicle_Break_SPC.astype(int)
Add_Vehicle_Break_SPC_corr=Add_Vehicle_Break_SPC.corr()
Add_Vehicle_Break_SPC_corr=Add_Vehicle_Break_SPC_corr.iloc[-1]
Add_Vehicle_Break_SPC_corr.sort_values(ascending = False)

Subscription_Premium_Cancel    1.000000
Add_Vehicle_Break             -0.911105
Name: Subscription_Premium_Cancel, dtype: float64

Немає кореляції між тим, що клієнт додав паузу у русі автомобіля та тим, що він скасував підписку

- Add_Payment_Method_Failed

In [53]:
Add_Payment_Method_Failed.value_counts()

Add_Payment_Method_Failed                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       
{"Error Code":"ERROR-16010","Error Message":"Card is debit but credit is required.","Payment Method":"Credit Card","Tokenized Pay":""}                                                                                                                                                                                                                                                                                                                                 

In [54]:
# Якщо у клієнта сталася помилка з додаванням методу оплати - то 1, якщо ні - то 0
Add_Payment_Method_Failed.loc[~Add_Payment_Method_Failed['Add_Payment_Method_Failed'].isnull()] = 1

In [55]:
Add_Payment_Method_Failed_SPC = pd.concat([Add_Payment_Method_Failed, Subscription_Premium_Cancel], axis=1, join='outer')
Add_Payment_Method_Failed_SPC = Add_Payment_Method_Failed_SPC.fillna(0)
Add_Payment_Method_Failed_SPC=Add_Payment_Method_Failed_SPC.astype(int)
Add_Payment_Method_Failed_SPC_corr=Add_Payment_Method_Failed_SPC.corr()
Add_Payment_Method_Failed_SPC_corr=Add_Payment_Method_Failed_SPC_corr.iloc[-1]
Add_Payment_Method_Failed_SPC_corr.sort_values(ascending = False)

Subscription_Premium_Cancel    1.0
Add_Payment_Method_Failed     -1.0
Name: Subscription_Premium_Cancel, dtype: float64

Бачимо, що взагалі немає такого клієнта, в якого сталася помилка з додаванням методу оплати, щоб він скасував підписку

- Transaction_Refund

In [56]:
Transaction_Refund.value_counts()

Transaction_Refund                                         
{"Id":"052ba99232fe3c50eef480d6fb6f362b","Amount":"-50"}       1
{"Id":"988623b2afe555a1ee3f51bacbcbf064","Amount":"-8.14"}     1
{"Id":"ae886bef41f1bee1fcb9b9d21453e1bd","Amount":"-50"}       1
{"Id":"abb700649f62a3d92ddcdf33e98862e1","Amount":"-10"}       1
{"Id":"aa01b6224e43702c3f89fec0f893bd0a","Amount":"-16.62"}    1
                                                              ..
{"Id":"417fb1dfecc15925902300432b3ae4ef","Amount":"-20"}       1
{"Id":"3f13bdc53ae75f950f141d454534591b","Amount":"-20"}       1
{"Id":"3c8b2b7ed2b8063f51eb31d70ed00e26","Amount":"-20"}       1
{"Id":"3857f1c2e0c0df4b42c3d0cb7839908e","Amount":"-20"}       1
{"Id":"ffa17017e78d166c92265aa97d515dd3","Amount":"-50"}       1
Length: 78, dtype: int64

In [57]:
# Якщо клієнту повернули транзакцію - то 1, якщо ні - то 0
Transaction_Refund.loc[~Transaction_Refund['Transaction_Refund'].isnull()] = 1

In [58]:
Transaction_Refund_SPC = pd.concat([Transaction_Refund, Subscription_Premium_Cancel], axis=1, join='outer')
Transaction_Refund_SPC = Transaction_Refund_SPC.fillna(0)
Transaction_Refund_SPC=Transaction_Refund_SPC.astype(int)
Transaction_Refund_SPC_corr=Transaction_Refund_SPC.corr()
Transaction_Refund_SPC_corr=Transaction_Refund_SPC_corr.iloc[-1]
Transaction_Refund_SPC_corr.sort_values(ascending = False)

Subscription_Premium_Cancel    1.000000
Transaction_Refund            -0.788227
Name: Subscription_Premium_Cancel, dtype: float64

Немає корелеції між тим, що клієнту повернули гроші та тим, що він скасував підписку

- event_platform

In [59]:
df.event_platform.unique()

array(['android', 'ios', 'outofband'], dtype=object)

In [60]:
platforms=['android', 'ios', 'outofband']


Platformss=pd.DataFrame()
for platform in platforms:
     Platformss[platform] = df['event_platform'].str.find(platform)



Platformss = Platformss.replace(0,1) 
Platformss = Platformss.replace(-1,0) 

Platformss.drop_duplicates(keep='first')
Platformss = Platformss.groupby(by=["userid"]).sum()
Platformss = Platformss.fillna(0)


Platformss_SPC = pd.concat([Platformss, Subscription_Premium_Cancel], axis=1, join='outer')
Platformss_SPC = Platformss_SPC.fillna(0)
Platformss_SPC=Platformss_SPC.astype(int)
Platformss_SPC_corr=Platformss_SPC.corr()
Platformss_SPC_corr=Platformss_SPC_corr.iloc[-1]
Platformss_SPC_corr.sort_values(ascending = False)

Subscription_Premium_Cancel    1.000000
outofband                      0.295000
ios                            0.106034
android                        0.064420
Name: Subscription_Premium_Cancel, dtype: float64

Найбільша кореляція до скасування підписки - коли в клієнта outofband , кореляція дорівнює 0.295000

Кореляція між штатом клієнта та скасуванням підписки 

In [61]:
states = [ 'AK', 'AL', 'AR', 'AZ', 'CA', 'CO', 'CT', 'DC', 'DE', 'FL', 'GA',
           'HI', 'IA', 'ID', 'IL', 'IN', 'KS', 'KY', 'LA', 'MA', 'MD', 'ME',
           'MI', 'MN', 'MO', 'MS', 'MT', 'NC', 'NE', 'NH', 'NJ', 'NM',
           'NV', 'NY', 'OH', 'OK', 'OR', 'PA', 'RI', 'SC', 'SD', 'TN', 'TX',
           'UT', 'VA', 'VT', 'WA', 'WI', 'WV']


States=pd.DataFrame()
for state in states:
     States[state] = df['user_state'].str.find(state)

States = States.replace(0,1) 
States = States.replace(-1,0) 


States=States.drop_duplicates()
States = States.fillna(0)
States=States.astype(int)



States_SPC = pd.concat([States, Subscription_Premium_Cancel], axis=1, join='outer')
States_SPC = States_SPC.fillna(0)
States_SPC=States_SPC.astype(int)
States_SPC_corr=States_SPC.corr()
States_SPC_corr=States_SPC_corr.iloc[-1]
States_SPC_corr.sort_values(ascending = False)

Subscription_Premium_Cancel    1.000000
TN                             0.159586
KS                            -0.084678
VT                            -0.084678
UT                            -0.084678
NM                            -0.084678
AL                            -0.084678
ME                            -0.084678
LA                            -0.084678
KY                            -0.084678
AK                            -0.084678
MS                            -0.084678
MT                            -0.084678
WV                            -0.084678
WI                            -0.084678
VA                            -0.084678
SC                            -0.084678
PA                            -0.084678
OR                            -0.084678
OK                            -0.084678
OH                            -0.084678
NY                            -0.084678
NV                            -0.084678
IA                            -0.084678
NH                            -0.084678


Найбільша кореляція між клієнтами зі штату Tеннессі та скасуванням підписки, вона дорівнює 0.159586 .

- Івенти без параметрів

Якщо клієнт робив щось з цих івентів - то 1, якщо ні - то 0, незважаємо на параметри
- Transaction Refund 
- Order 
- Account Setup Skip 
- Signup Success 
- Add Vehicle Success 
- Add Vehicle Break
- Add Payment Method Sucess 
- Add Payment Method Failed
- Email Confirmation Success
- Chat conversation Opened
- Sign Out 
- Account History Transaction Details
- Wallet Opened
- Subscription Premium 
- Subscription Premium Cancel
- Calculator view 

In [62]:
url = 'https://github.com/Nahorskyi/hackathon_INT20H_test/blob/main/int20h-ds-test-dataset.csv?raw=true'
df = pd.read_csv(url)
df=df.set_index('userid')
df.pop('event_created_date')
df=df.drop_duplicates(keep='first')

In [63]:
Transaction_Refund = df[df['event_name'] == 'Transaction Refund'][['event_attributes']] 
Transaction_Refund.rename(columns = {'event_attributes': 'Transaction_Refund'}, inplace=True)
Transaction_Refund=Transaction_Refund.groupby(by=["userid"]).sum()
Transaction_Refund.loc[~Transaction_Refund['Transaction_Refund'].isnull()] = 1
Transaction_Refund.Transaction_Refund.value_counts ()

1    78
Name: Transaction_Refund, dtype: int64

In [64]:
Order = df[df['event_name'] == 'Order'][['event_attributes']] 
Order.rename(columns = {'event_attributes':'Order'} , inplace = True ) 
Order=Order.groupby(by=["userid"]).sum()
Order.loc[~Order['Order'].isnull()] = 1
Order.Order.value_counts ()

1    39
Name: Order, dtype: int64

In [65]:
Account_Setup_Profile_Skip = df[df['event_name'] == 'Account Setup Profile Skip'][['event_attributes']] 
Account_Setup_Profile_Skip.rename(columns = {'event_attributes': 'Account_Setup_Profile_Skip'}, inplace=True)
Account_Setup_Profile_Skip=Account_Setup_Profile_Skip.groupby(by=["userid"]).sum()
Account_Setup_Profile_Skip.loc[~Account_Setup_Profile_Skip['Account_Setup_Profile_Skip'].isnull()] = 1
Account_Setup_Profile_Skip.Account_Setup_Profile_Skip.value_counts ()

1    188
Name: Account_Setup_Profile_Skip, dtype: int64

In [66]:
Sign_Up_Success = df[df['event_name'] == 'Sign Up Success'][['event_attributes']] 
Sign_Up_Success.rename(columns = {'event_attributes': 'Sign_Up_Success'}, inplace=True)
Sign_Up_Success=Sign_Up_Success.groupby(by=["userid"]).sum()
Sign_Up_Success.loc[~Sign_Up_Success['Sign_Up_Success'].isnull()] = 1
Sign_Up_Success.Sign_Up_Success.value_counts ()

1    1460
Name: Sign_Up_Success, dtype: int64

In [67]:
Add_Vehicle_Success = df[df['event_name'] == 'Add Vehicle Success'][['event_attributes']] 
Add_Vehicle_Success.rename(columns = {'event_attributes': 'Add_Vehicle_Success'}, inplace=True)
Add_Vehicle_Success=Add_Vehicle_Success.groupby(by=["userid"]).sum() 
Add_Vehicle_Success.loc[~Add_Vehicle_Success['Add_Vehicle_Success'].isnull()] = 1
Add_Vehicle_Success.Add_Vehicle_Success.value_counts ()

1    1790
Name: Add_Vehicle_Success, dtype: int64

In [68]:
Add_Vehicle_Break  = df[df['event_name'] == 'Add Vehicle Break'][['event_attributes']] 
Add_Vehicle_Break.rename(columns = {'event_attributes': 'Add_Vehicle_Break'}, inplace=True)
Add_Vehicle_Break=Add_Vehicle_Break.groupby(by=["userid"]).sum()
Add_Vehicle_Break.loc[~Add_Vehicle_Break['Add_Vehicle_Break'].isnull()] = 1
Add_Vehicle_Break.Add_Vehicle_Break.value_counts ()
# True False

1    211
Name: Add_Vehicle_Break, dtype: int64

In [69]:
Add_Payment_Method_Success = df[df['event_name'] == 'Add Payment Method Success'][['event_attributes']] 
Add_Payment_Method_Success.rename(columns = {'event_attributes': 'Add_Payment_Method_Success'}, inplace=True) 
Add_Payment_Method_Success=Add_Payment_Method_Success.groupby(by=["userid"]).sum()
Add_Payment_Method_Success.loc[~Add_Payment_Method_Success['Add_Payment_Method_Success'].isnull()] = 1
Add_Payment_Method_Success.Add_Payment_Method_Success.value_counts ()

1    709
Name: Add_Payment_Method_Success, dtype: int64

In [70]:
Add_Payment_Method_Failed   = df[df['event_name'] == 'Add Payment Method Failed'][['event_attributes']] 
Add_Payment_Method_Failed.rename(columns = {'event_attributes': 'Add_Payment_Method_Failed'}, inplace=True)
Add_Payment_Method_Failed =Add_Payment_Method_Failed.groupby(by=["userid"]).sum()
Add_Payment_Method_Failed.loc[~Add_Payment_Method_Failed['Add_Payment_Method_Failed'].isnull()] = 1
Add_Payment_Method_Failed.Add_Payment_Method_Failed.value_counts ()

1    210
Name: Add_Payment_Method_Failed, dtype: int64

In [71]:
Email_Confirmation_Success = df[df['event_name'] == 'Email Confirmation Success'][['event_attributes']] 
Email_Confirmation_Success.rename(columns = {'event_attributes': 'Email_Confirmation_Success'}, inplace=True)
Email_Confirmation_Success=Email_Confirmation_Success.groupby(by=["userid"]).sum()
Email_Confirmation_Success.loc[~Email_Confirmation_Success['Email_Confirmation_Success'].isnull()] = 1
Email_Confirmation_Success.Email_Confirmation_Success.value_counts ()

1    361
Name: Email_Confirmation_Success, dtype: int64

In [72]:
Chat_Conversation_Opened   = df[df['event_name'] == 'Chat Conversation Opened'][['event_attributes']] 
Chat_Conversation_Opened.rename(columns = {'event_attributes': 'Chat_Conversation_Opened'}, inplace=True) 
Chat_Conversation_Opened=Chat_Conversation_Opened.groupby(by=["userid"]).sum()
Chat_Conversation_Opened.loc[~Chat_Conversation_Opened['Chat_Conversation_Opened'].isnull()] = 1
Chat_Conversation_Opened.Chat_Conversation_Opened.value_counts ()

1    320
Name: Chat_Conversation_Opened, dtype: int64

In [73]:
Sign_Out = df[df['event_name'] == 'Sign Out'][['event_attributes']] 
Sign_Out.rename(columns = {'event_attributes': 'Sign_Out'}, inplace=True)
Sign_Out =Sign_Out.groupby(by=["userid"]).sum()
Sign_Out .loc[~Sign_Out ['Sign_Out'].isnull()] = 1
Sign_Out.Sign_Out .value_counts ()

1    228
Name: Sign_Out, dtype: int64

In [74]:
Account_History_Transaction_Details = df[df['event_name'] == 'Account History Transaction Details'][['event_attributes']] 
Account_History_Transaction_Details.rename(columns = {'event_attributes': 'Account_History_Transaction_Details'}, inplace=True) 
Account_History_Transaction_Details=Account_History_Transaction_Details.groupby(by=["userid"]).sum()
Account_History_Transaction_Details.loc[~Account_History_Transaction_Details['Account_History_Transaction_Details'].isnull()] = 1
Account_History_Transaction_Details.Account_History_Transaction_Details.value_counts ()

1    319
Name: Account_History_Transaction_Details, dtype: int64

In [75]:
Wallet_Opened   = df[df['event_name'] == 'Wallet Opened'][['event_attributes']] 
Wallet_Opened.rename(columns = {'event_attributes': 'Wallet_Opened'}, inplace=True) 
Wallet_Opened=Wallet_Opened.groupby(by=["userid"]).sum()
Wallet_Opened.loc[~Wallet_Opened['Wallet_Opened'].isnull()] = 1
Wallet_Opened.Wallet_Opened.value_counts ()
# 

1    267
Name: Wallet_Opened, dtype: int64

In [76]:
Subscription_Premium = df[df['event_name'] == 'Subscription Premium'][['event_attributes']] 
Subscription_Premium.rename(columns = {'event_attributes': 'Subscription_Premium'}, inplace=True)
Subscription_Premium=Subscription_Premium.groupby(by=["userid"]).sum()
Subscription_Premium.loc[~Subscription_Premium['Subscription_Premium'].isnull()] = 1
Subscription_Premium.Subscription_Premium.value_counts ()

1    356
Name: Subscription_Premium, dtype: int64

In [77]:
Subscription_Premium_Cancel   = df[df['event_name'] == 'Subscription Premium Cancel'][['event_attributes']] 
Subscription_Premium_Cancel.rename(columns = {'event_attributes': 'Subscription_Premium_Cancel'}, inplace=True)
Subscription_Premium_Cancel=Subscription_Premium_Cancel.groupby(by=["userid"]).sum()
Subscription_Premium_Cancel.loc[~Subscription_Premium_Cancel['Subscription_Premium_Cancel'].isnull()] = 1
Subscription_Premium_Cancel.Subscription_Premium_Cancel.value_counts ()
# True False

1    26
Name: Subscription_Premium_Cancel, dtype: int64

In [78]:
Calculator_View = df[df['event_name'] == 'Calculator View'][['event_attributes']] 
Calculator_View.rename(columns = {'event_attributes': 'Calculator_View'}, inplace=True)
Calculator_View=Calculator_View.groupby(by=["userid"]).sum()
Calculator_View.loc[~Calculator_View['Calculator_View'].isnull()] = 1
Calculator_View.Calculator_View.value_counts ()
# True False

1    233
Name: Calculator_View, dtype: int64

In [79]:
res_int = pd.concat([Sign_Up_Success,Order, Add_Vehicle_Success,Account_History_Transaction_Details, Chat_Conversation_Opened, Wallet_Opened, Add_Payment_Method_Success, Email_Confirmation_Success, Subscription_Premium, Calculator_View, Sign_Out, Account_Setup_Profile_Skip, Add_Vehicle_Break, Add_Payment_Method_Failed,  Transaction_Refund,Subscription_Premium_Cancel], axis=1, join='outer')

res_int = res_int.replace(['{}'],'1')
res_int = res_int.fillna(0)

corr_res_int=res_int.corr()
corr_res_int=corr_res_int.iloc[-1]
corr_res_int.sort_values(ascending = False)

Subscription_Premium_Cancel            1.000000
Wallet_Opened                          0.234693
Order                                  0.179072
Transaction_Refund                     0.164547
Subscription_Premium                   0.141731
Chat_Conversation_Opened               0.128915
Account_History_Transaction_Details    0.117476
Add_Payment_Method_Success             0.088945
Sign_Out                               0.052705
Calculator_View                        0.051557
Email_Confirmation_Success             0.040324
Add_Vehicle_Break                      0.028626
Sign_Up_Success                        0.017651
Account_Setup_Profile_Skip            -0.011256
Add_Vehicle_Success                   -0.021230
Add_Payment_Method_Failed             -0.027743
Name: Subscription_Premium_Cancel, dtype: float64

Найбільша кореляція між Subscription_Premium_Cancel та Wallet_Opened(коли клієнт відкрив гаманець), яка становить 0.234693.

Найменша колеляція між Subscription_Premium_Cancel та Add_Payment_Method_Failed, яка становить -0.027743

- Івенти з параметрами та характеристика клієнта

In [80]:
res_int_par = pd.concat([Platformss, States, Sign_Up_Success_all,Order_all,Add_Vehicle_Success_all,Account_History_Transaction_Details,Chat_Conversation_Opened,Wallet_Opened_all,Add_Payment_Method_Success,Email_Confirmation_Success_all,Subscription_Premium,Calculator_View_all,Sign_Out,Account_Setup_Profile_Skip,Add_Vehicle_Break,Add_Payment_Method_Failed,Transaction_Refund,Subscription_Premium_Cancel], axis=1, join='outer')

res_int_par = res_int_par.replace(['{}'],'1')
res_int_par = res_int_par.fillna(0)


res_int_par_corr=res_int_par.corr()
res_int_par_corr=res_int_par_corr.iloc[-1]
res_int_par_corr.sort_values(ascending = False)

Subscription_Premium_Cancel    1.000000
outofband                      0.295000
Wallet_Opened_1                0.224346
Order_True                     0.219163
TN                             0.195250
                                 ...   
Toyota                        -0.020644
Sign_Up_Success_Facebook      -0.023667
Nissan                        -0.024129
Ford                          -0.026200
Add_Payment_Method_Failed     -0.027192
Name: Subscription_Premium_Cancel, Length: 105, dtype: float64

Бачимо, що найбільша кореляція між скасуванням підписки та outofband (стосується платформи пристрою) становить 0.295000

Найменша кореляція між скасуванням підписки та Add_Payment_Method_Failed становить -0.027743